src/imitation/algorithms/bc_torch.py

```diff
@@ -0,0 +1,275 @@
+        and the current progress remaining (starts from 1 and ends at 0).
+
+        Args:
+            optimizers: An optimizer or a list of optimizers to update.
+        """
+        if not isinstance(optimizers, list):
+            optimizers = [optimizers]
+        for optimizer in optimizers:
+            update_learning_rate(
+                optimizer, self.lr_schedule(self._current_progress_remaining)
+            )
+
+    def _calculate_loss(self, obs, acts) -> Tuple[th.Tensor, Dict[str, float]]:
+        """
+        Calculate the supervised learning loss used to train the behavioral clone.
+
+        Args:
+            obs: The observations seen by the expert.
+            acts: The actions taken by the expert.
+
+        Returns:
+            loss: The supervised learning loss for the behavioral clone to optimize.
+            stats_dict: Statistics about the learning process to be logged.
+
+        """
+        _, log_prob, entropy = self.policy.evaluate_actions(obs, acts)
+        prob_true_act = th.exp(log_prob).mean()
+        log_prob = log_prob.mean()
+        entropy = entropy.mean()
+
+        ent_loss = -self.ent_weight * entropy
+        neglogp = -log_prob
+        loss = neglogp + ent_loss
+
+        l2_norms = [th.norm(w, 2) ** 2 for w in self.policy.parameters()]
+        l2_norm = sum(l2_norms) / 2  # divide by 2 to cancel with gradient of square
+        l2_loss = self.l2_weight * l2_norm
+
+        stats_dict = dict(
+            neglogp=neglogp.item(),
+            loss=loss.item(),
+            entropy=entropy.item(),
+            ent_loss=ent_loss.item(),
+            prob_true_act=prob_true_act.item(),
+            l2_norm=l2_norm.item(),
+            l2_loss=l2_loss.item(),
+        )
+
+        return loss, stats_dict
+
+    def train(
+        self,
+        n_epochs: int = 100,
+        *,
+        on_epoch_end: Callable[[dict], None] = None,
+        log_interval: int = 100,
+    ):
+        """Train with supervised learning for some number of epochs.
+
+        Here an 'epoch' is just a complete pass through the expert transition
+        dataset.
+
+        Args:
+          n_epochs: number of complete passes made through dataset.
+          on_epoch_end: optional callback to run at
+            the end of each epoch. Will receive all locals from this function as
+            dictionary argument (!!).
+          log_interval: log stats after every log_interval batches
+        """
+        assert self.batch_size >= 1
+        total_samples = n_epochs * self.expert_dataset.size()
+        total_batches = np.ceil(total_samples / self.batch_size)
+        samples_so_far = 0
+        batch_num = 0
+        for epoch_num in trange(n_epochs, desc="BC epoch"):
+            while samples_so_far < (epoch_num + 1) * self.expert_dataset.size():
+                self._update_current_progress_remaining(batch_num, total_batches)
+                self._update_learning_rate(self.policy.optimizer)
```

Comment:
I think `policy.optimizer` is always a single optimizer. I originally had `_update_learning_rate` handling a list of optimizers because I modeled the method after `stable_baselines3.common.base_class.BaseAlgorithm._update_learning_rate`. In the most recent commit, however, I removed the definition of `_update_learning_rate` from the bc class and am instead just directly calling `stable_baselines3.common.base_class.BaseAlgorithm._update_learning_rate`.